File: backend/src/main/java/com/corporate/rides/service/RideService.java
package com.corporate.rides.service;

import com.corporate.rides.dto.*;
import com.corporate.rides.enums.RideStatus;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

public interface RideService {
    RideResponseDto createRide(CreateRideRequestDto request);
    List<RideResponseDto> getEmployeeRides();
    RideResponseDto getRideById(UUID rideId);
    RideResponseDto cancelRide(UUID rideId, CancelRideRequestDto request);

    // Feature 4 — Ride Scheduling & Manager Request Management Methods
    List<RideResponseDto> getSchedulableRides();
    RideResponseDto approveRide(UUID rideId);
    RideResponseDto rejectRideRequest(UUID rideId, RejectRideRequestDto request);
    RideResponseDto scheduleRide(UUID rideId, ScheduleRideRequestDto request);
    RideResponseDto rescheduleRide(UUID rideId, RescheduleRideRequestDto request);
    List<RideResponseDto> getScheduledRides(String search, LocalDate bookingDate, RideStatus status);

    // Feature 5 — Driver & Vehicle Assignment Methods
    List<RideResponseDto> getPendingAssignmentRides();
    AssignmentOptionsResponseDto getAssignmentOptions(UUID rideId);
    RideResponseDto assignRideResources(UUID rideId, AssignRideRequestDto request);
    RideResponseDto replaceRideAssignment(UUID rideId, AssignRideRequestDto request);
    RideResponseDto unassignRideResources(UUID rideId);

    // Feature 6 & 11 — Real-Time Ride Tracking & Trip Completion Methods
    RideResponseDto startTrip(UUID rideId);
    LocationResponseDto updateLocation(UUID rideId, LocationUpdateDto dto);
    LocationResponseDto getLatestLocation(UUID rideId);
    List<LocationResponseDto> getLocationHistory(UUID rideId);
    RideResponseDto completeTrip(UUID rideId);
    RideResponseDto completeTrip(UUID rideId, CompleteRideRequestDto request);
    List<RideResponseDto> getActiveTrips();
    List<RideResponseDto> getDriverAssignedTrips();
    List<RideResponseDto> getCompletedTrips(String search, UUID driverId, LocalDate fromDate, LocalDate toDate);

    // Feature 7 — Driver Operations Methods
    RideResponseDto acceptRideAssignment(UUID rideId);
    RideResponseDto rejectRideAssignment(UUID rideId, RejectRideRequestDto request);
    RideResponseDto verifyEmployeeForRide(UUID rideId, EmployeeVerificationRequestDto request);
    List<RideResponseDto> getDriverTodayRides();
    List<RideResponseDto> getDriverRideHistory(LocalDate from, LocalDate to, RideStatus status);
}

File: frontend/src/services/rideService.js
import { apiFetch } from './api';

export const rideService = {
  createRide: async (rideData) => {
    const response = await apiFetch('/rides', {
      method: 'POST',
      body: JSON.stringify(rideData),
    });
    return response.data;
  },

  getEmployeeRides: async () => {
    const response = await apiFetch('/rides', {
      method: 'GET',
    });
    return response.data;
  },

  getRideById: async (id) => {
    const response = await apiFetch(`/rides/${id}`, {
      method: 'GET',
    });
    return response.data;
  },

  cancelRide: async (id, cancellationReason) => {
    const response = await apiFetch(`/rides/${id}/cancel`, {
      method: 'POST',
      body: JSON.stringify({ cancellationReason }),
    });
    return response.data;
  },

  // Feature 4 — Ride Scheduling & Manager Request Management API Methods
  getSchedulableRides: async () => {
    const response = await apiFetch('/rides/schedulable', {
      method: 'GET',
    });
    return response.data;
  },

  approveRide: async (id) => {
    const response = await apiFetch(`/rides/${id}/approve`, {
      method: 'POST',
    });
    return response.data;
  },

  rejectRideRequest: async (id, reason, notes = '') => {
    const response = await apiFetch(`/rides/${id}/reject-request`, {
      method: 'POST',
      body: JSON.stringify({ reason, notes }),
    });
    return response.data;
  },

  scheduleRide: async (id, scheduleData) => {
    const response = await apiFetch(`/rides/${id}/schedule`, {
      method: 'POST',
      body: JSON.stringify(scheduleData),
    });
    return response.data;
  },

  rescheduleRide: async (id, rescheduleData) => {
    const response = await apiFetch(`/rides/${id}/reschedule`, {
      method: 'PATCH',
      body: JSON.stringify(rescheduleData),
    });
    return response.data;
  },

  getScheduledRides: async (params = {}) => {
    const queryParams = new URLSearchParams();
    if (params.search) queryParams.append('search', params.search);
    if (params.bookingDate) queryParams.append('bookingDate', params.bookingDate);
    if (params.status && params.status !== 'ALL') queryParams.append('status', params.status);

    const queryString = queryParams.toString();
    const endpoint = `/rides/scheduled${queryString ? `?${queryString}` : ''}`;

    const response = await apiFetch(endpoint, {
      method: 'GET',
    });
    return response.data;
  },

  // Feature 5 — Driver & Vehicle Assignment API Methods
  getPendingAssignmentRides: async () => {
    const response = await apiFetch('/rides/assignment-pending', {
      method: 'GET',
    });
    return response.data;
  },

  getAssignmentOptions: async (id) => {
    const response = await apiFetch(`/rides/${id}/assignment-options`, {
      method: 'GET',
    });
    return response.data;
  },

  assignRideResources: async (id, assignmentData) => {
    const response = await apiFetch(`/rides/${id}/assign`, {
      method: 'POST',
      body: JSON.stringify(assignmentData),
    });
    return response.data;
  },

  assignDriverAndVehicle: async (id, assignmentData) => {
    const response = await apiFetch(`/rides/${id}/assign`, {
      method: 'POST',
      body: JSON.stringify(assignmentData),
    });
    return response.data;
  },

  replaceRideAssignment: async (id, assignmentData) => {
    const response = await apiFetch(`/rides/${id}/assignment`, {
      method: 'PATCH',
      body: JSON.stringify(assignmentData),
    });
    return response.data;
  },

  unassignRideResources: async (id) => {
    const response = await apiFetch(`/rides/${id}/assignment`, {
      method: 'DELETE',
    });
    return response.data;
  },

  unassignRide: async (id) => {
    const response = await apiFetch(`/rides/${id}/assignment`, {
      method: 'DELETE',
    });
    return response.data;
  },

  // Feature 6 — Real-Time Tracking & Monitoring API Methods
  startTrip: async (id) => {
    const response = await apiFetch(`/rides/${id}/start`, {
      method: 'POST',
    });
    return response.data;
  },

  updateLocation: async (id, locationData) => {
    const response = await apiFetch(`/rides/${id}/location`, {
      method: 'POST',
      body: JSON.stringify(locationData),
    });
    return response.data;
  },

  getLatestLocation: async (id) => {
    const response = await apiFetch(`/rides/${id}/location`, {
      method: 'GET',
    });
    return response.data;
  },

  getLocationHistory: async (id) => {
    const response = await apiFetch(`/rides/${id}/location/history`, {
      method: 'GET',
    });
    return response.data;
  },

  completeTrip: async (id) => {
    const response = await apiFetch(`/rides/${id}/complete`, {
      method: 'POST',
    });
    return response.data;
  },

  getActiveTrips: async () => {
    const response = await apiFetch('/rides/active', {
      method: 'GET',
    });
    return response.data;
  },

  getDriverAssignedTrips: async () => {
    const response = await apiFetch('/rides/driver-assigned', {
      method: 'GET',
    });
    return response.data;
  },

  // Feature 7 — Driver Operations API Methods
  acceptRide: async (id) => {
    const response = await apiFetch(`/rides/${id}/accept`, {
      method: 'POST',
    });
    return response.data;
  },

  rejectRide: async (id, reason, notes = '') => {
    const response = await apiFetch(`/rides/${id}/reject`, {
      method: 'POST',
      body: JSON.stringify({ reason, notes }),
    });
    return response.data;
  },

  verifyEmployee: async (id, employeeIdentifier, verificationMethod = 'BADGE_OR_EMAIL') => {
    const response = await apiFetch(`/rides/${id}/verify-employee`, {
      method: 'POST',
      body: JSON.stringify({ employeeIdentifier, verificationMethod }),
    });
    return response.data;
  },

  getDriverTodayRides: async () => {
    const response = await apiFetch('/rides/driver/today', {
      method: 'GET',
    });
    return response.data;
  },

  getDriverHistory: async (params = {}) => {
    const queryParams = new URLSearchParams();
    if (params.from) queryParams.append('from', params.from);
    if (params.to) queryParams.append('to', params.to);
    if (params.status && params.status !== 'ALL') queryParams.append('status', params.status);

    const queryString = queryParams.toString();
    const endpoint = `/rides/driver/history${queryString ? `?${queryString}` : ''}`;

    const response = await apiFetch(endpoint, {
      method: 'GET',
    });
    return response.data;
  },
};
